### ISSUE TITLE
Incorrect R1/R2 penalty because of second order gradients

### ISSUE THREAD

**Issue by HasnainRaz** [brownvc/R3GAN#20]
I saw the implementation of the R1/R2 penalties here: https://github.com/brownvc/R3GAN/blob/main/R3GAN/Trainer.py#L35

When we call this, we get gradients propagated back into the input to the Discriminator. Then when we call backward again:
https://github.com/brownvc/R3GAN/blob/main/R3GAN/Trainer.py#L42, this populates gradients again in the input.
This creates second order gradients.

This is incorrect right? as the intention of the original penalty is to not be able to minimize it through second order gradients.

Adding a .detach() after calculating the penalty will resolve this, since the .backward call will stop the propagation of the penalty gradients a second time and likely lower the memory usage (since second order grads are not computed)

I'm not sure if this bug affects the findings presented in the research paper itself? I have been trying to implement this penalty for other use cases and do not observe the same nice convergence patterns as reported in the paper. And I'm just wondering if it is because of this bug?

Let me know your thoughts.

**Comment by nick-bvc** [comment]
Our implementation is correct.
If you added `.detach()` after the penalty terms, your implementation is incorrect as the penalty terms would be a constant to the optimization.
